docs/qos_enforcement.md

```diff
@@ -0,0 +1,254 @@
+4. Assignment of QoS profiles to APN entities (to denote what QoS levels are supported by the APN)
+5. Assignment of APN entities to subscribers (to denote which APNs a subscriber has access to)
+6. Assignment of ApnServiceType entities to gateways
+
+**Federated LTE**
+
+In federated use cases, we have slightly different requirements. APN entities 
+do not need to be configured as they are provided by the federation network.
+
+We also wish to support cases where an HSS is not provided, and our 
+`subscriberdb` service is used, but where the PCRF is federated, 
+and so our `policydb` service is disabled.
+
+This adds the following requirement:
+
+7. Disabled configuration of APN entities for federated LTE networks
+
+## How We Will Change Magma
+
+### API Endpoints
+
+To enable configuration of QoS profiles and APNs, we will support the following REST endpoints:
+
+```
+QoS Profiles
+
+GET     /lte/{network_id}/qos_profiles
+POST    /lte/{network_id}/qos_profiles
+
+GET     /lte/{network_id}/qos_profiles/{qos_name}
+PUT     /lte/{network_id}/qos_profiles/{qos_name}
+DELETE  /lte/{network_id}/qos_profiles/{qos_name}
+
+GET     /feg_lte/{network_id}/qos_profiles
+POST    /feg_lte/{network_id}/qos_profiles
+
+GET     /feg_lte/{network_id}/qos_profiles/{qos_name}
+PUT     /feg_lte/{network_id}/qos_profiles/{qos_name}
+DELETE  /feg_lte/{network_id}/qos_profiles/{qos_name}
+
+
+
+APN Profiles
+
+GET     /lte/{network_id}/apns
+POST    /lte/{network_id}/apns
+
+GET     /lte/{network_id}/apns/{apn_name}
+PUT     /lte/{network_id}/apns/{apn_name}
+DELETE  /lte/{network_id}/apns/{apn_name}
+
+
+
+APN Service Type
+
+GET     /lte/{network_id}/gateways/{gateway_id}/service_types
+POST    /lte/{network_id}/gateways/{gateway_id}/service_types
+
+GET     /lte/{network_id}/gateways/{gateway_id}/service_types/{service_type}
+PUT     /lte/{network_id}/gateways/{gateway_id}/service_types/{service_type}
+DELETE  /lte/{network_id}/gateways/{gateway_id}/service_types/{service_type}
+```
+
+
+
+### API Entity Definitions
+
+Entity definitions for QoS, APN, and subscribers will change. The final state is summarized here:
+
+```
+qos_profile:
+  id: string
+  class_id: enum[0,255]
+  max_req_bw_dl: uint32
+  max_req_bw_ul: uint32
+  # Guaranteed bit rate
+  # Not enforced by Magma yet
+  gbr:
+    gbr_ul: uint32
+    gbr_dl: uint32
+  # Allocation and retention priority
+  # Not enforced by Magma yet
+  arp:
+    arp_priority_level: uint32
+    arp_preemption_capability: bool
+    arp_preemption_vulnerability: bool
+ # Aggregated Maximum bit rate
+ # for non-GBR bearers, this limits the aggregate bit rate
+ # Enforce this on pipelined
+ apn_ambr:
+   apn_ambr_dl: uint32
+   apn_ambr_ul: uint32
+```
+
+As an initial goal, we would like to support apn_ambr.
+
+```
+apn_configuration:
+  supported_qos_profiles: array[qos_profile]
+  # service_type should match the name of a specified ApnServiceType
+  service_type: string
```

Comment:
change `service_type` to `apn_resource_config`

Reply:
Actually this does not make sense to me to define this as part of APN configuration. This is an AGW level resource.

Reply:
strike my comment, actually you are doing it AGW to service and APN to service, with service -> definition.